Configurable email sending error handling

taleinat opened the pull request:
This is my continuation of the work done by @volksman on his `error_handler` branch, see #92.

I've added documentation and improved the default error handler.

taleinat commented:
@spookylukey, please at least take a look at this?

spookylukey commented on mailer/engine.py:
```diff
@@ -112,6 +113,28 @@ def release_lock(lock):
+    )):
+        # TODO: Avoid endless retrying sending of messages on unrecognized
+        #       types of errors.
+        pass
```
This code at the moment does nothing (`pass` just means it will carry on to the line below). Did you mean `raise`? I think there might be a problem using a bare `raise` here because the code is in a different scope from where the exception was caught, so it might need to use `raise err`.

As written, this code catches all exceptions and makes the messages deferred. This changes the behaviour from before - the previous error handler only caught recognized errors, others propagated.

The previous behaviour had some good features that we don't want to break. In particular, in the case of transient network errors SMTPConnectError the previous behaviour was better.

But there are some changes we do want to make to the old behaviour.

I think this comment gets it right: 
* https://github.com/pinax/django-mailer/issues/73#issuecomment-265561162 .
With some caveats:
* My comment here: https://github.com/pinax/django-mailer/issues/73#issuecomment-312021658
* Instead of 'cleanly abort send_all', I'm happy with re-raising the original (transient) error, which will then cause the `send_all` call to be aborted and the `manage.py` command to exit as an error, which is actually what we want.

I think it should be possible to add tests for all this, checking that the default error handler does all the right things.

volksman commented:
@taleinat and @spookylukey I think the best compromise here is to include both and let the users decide which way they want to enable.  I can see @spookylukey points but they don't concern me at all since I'm not using SMTP at all.  And that's the whole point is simply the ability to write my own handler for the service I do use.  So I think a simple catch all and defer and a raise and be aware handler should come out of the box and the end user will need to define it in settings (default being the raise and aware).

That said I don't have cycles to help at the moment.

taleinat commented:
I've removed the dependency on `import_string` which broke support with old versions of Django. I've also updated the testing configuration to test with all relevant versions of Python and Django to better check that they are all supported.

I'll take a look at more specific handling of different errors next.

taleinat commented:
@spookylukey, regarding the error handling, when deleting a message, how would you expect "create a MessageLog with the original data" to work? I suggest we define a new state in addition to "deferred", e.g. "failed", so the message would simply be kept in the Message table but no longer retried.

taleinat commented:
I'm no longer interested in this, so closing this PR.
